diffrax/_integrate.py

```diff
@@ -775,9 +775,38 @@ def _save_t1(subsaveat, save_state):
                 save_state = _save(tfinal, yfinal, args, subsaveat.fn, save_state)
         return save_state
 
-    save_state = jtu.tree_map(
-        _save_t1, saveat.subs, final_state.save_state, is_leaf=_is_subsaveat
+    def _save_ts(subsaveat: SubSaveAt, save_state: SaveState) -> SaveState:
+        if subsaveat.ts is not None:
+            ys = jtu.tree_map(
+                lambda y: jnp.stack([y] * len(save_state.ts)),
+                subsaveat.fn(t0, yfinal, args),
+            )
+            save_state = SaveState(
+                saveat_ts_index=len(save_state.ts),
+                ts=save_state.ts,
+                ys=ys,
+                save_index=len(save_state.ts),
+            )
+        return save_state
+
+    # if t0 == t1 then we don't enter the integration loop. In this case we have to
+    # manually update the saved ts and ys if we want to save at "intermediate"
+    # times specified by saveat.subs.ts
+    save_state = jax.lax.cond(
+        eqxi.unvmap_any(t0 == t1),
+        lambda __save_state: jax.lax.cond(
+            t0 == t1,
+            lambda _save_state: jtu.tree_map(
+                _save_ts, saveat.subs, _save_state, is_leaf=_is_subsaveat
+            ),
+            lambda _save_state: _save_state,
+            __save_state,
+        ),
+        lambda __save_state: __save_state,
+        final_state.save_state,
     )
+
+    save_state = jtu.tree_map(_save_t1, saveat.subs, save_state, is_leaf=_is_subsaveat)
```

Comment:
So this pretty much all LGTM! I can see that you're using the invariant `t0 <= ts < t1`. That's much neater than the unnecessary `where` check that I was suggesting!

I think my only concerns now are how this edge case interacts with a couple of other edge cases:
1. if both `t0 == t1` and the conditions of `_save_t1` are triggered, then I think the latter will actually attempt to write out-of-bounds?
2. if we have a boolean-returning `event` that triggers immediately (which is another reason to never enter the integration loop), then do we do the right thing? (Whether `t0 == t1` or `t0 != t1`.)

I think (1) at least might be solved by putting the new `_save_ts` after `_save_t1`, so that we essentially just overwrite the latter if we hit this case -- WDYT?

Reply:
(Oh also you'll see a force-push on this branch -- I updated `dev` and this was what gave a readable diff for me to review here. I realised after-the-fact that this means you'll now need to force-pull your own local branch... sorry about that!)

Reply:
You're absolutely right about point 1: fixed in most recent commit. 

I'm having a hard time seeing what might go wrong with such a boolean event that immediately triggers. In this case, we should still just fill in `y0` as we do now if `subsaveat.ts==True` and `t0==t1`. If `t0!=t1` then the `cond` doesn't get triggered anyways and this discussion is moot. Do you agree?

I'm also wondering about what we should be doing if `t0==t1` and `subsaveat.steps==True`. If it is, but `subsaveat.ts is None`, then we don't fill in `y0` for any of the steps. This feels like the right behavior. On the other hand, if `subsaveat.steps==True` and `subsaveat.ts is not None`, then we fill in `y0` for all values in `save_state`, including those for the "steps". Do you think this is the right behavior?

Reply:
(Also sorry for the long delay, been a crazy week!)

Reply:
Ok, updated to not update steps when `t0==t1` and `subsaveat.steps==True`. Edge cases FTW!

Reply:
I also added a test for this case